### ISSUE TITLE
Port twisted.protocols.basic to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6025]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6025|
|Type|enhancement|
|Created|2012-09-28 15:12:46Z|
|Branch|https://github.com/twisted/twisted/tree/basic-py3-6025-3|

`twisted.protocols.basic` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6025 6025
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_basic_py3_6025_3 branches/basic-py3-6025-3
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1348845166000000 1348845166000000
changetime__1349292691000000 1349292691000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35830]) Branching to 'basic-py3-6025'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

Ugh, for some reason the branch wasn't updated. Build running here:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/basic-py3-6025-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  Looks like you had some fun on this one.

> Ugh, for some reason the branch wasn't updated. Build running here:

When this happens, please follow [the directions](http://twistedmatrix.com/trac/wiki/CommitterCheckList).

  1. The `test-case-name` on `twisted/protocols/basic.py` needs to be updated
  1. The byte literals in the documentation for NetstringReceiver._LENGTH should be in C{}
  1. Lots of tests in the newly created `twisted.protocols.test.test_basic` don't follow the coding standard (misnamed, poorly written docstrings, use `fail` variation of assertion methods, etc).  Please file a ticket for fixing them.
  1. I guess my general thoughts regarding `_PY` is that it should be used at top-level module scope and not anywhere else.  Particularly the cases which aren't going to go away, like cases that deal with the differences between Python 2 str and Python 3 bytes, as in `LineOnlyReceiverTestCase.test_buffer`.  Using it to define two alternate implementations of a helper function seems tolerable (and a function for getting an iterator over bytes that yields each individual byte as a bytes object instead of an integer sounds like one that will be used more than once anyway - new `twisted.python.compat` API?), but constantly re-checking it after import time bothers me, both for performance reasons and because of the tedium it introduces for debugging and overall cognitive load of a section of code.  This approach would avoid worsening the Python 2 performance of the implementation of `sendString`, too.
    1. Other tests in this module that could use such a helper:
      1. LPTestCaseMixin.test_illegal
      1. NetstringReceiverTestCase.test_receiveNetstringPortions_3
      1. IntNTestCaseMixin.test_receive
      1. IntNTestCaseMixin.test_partial
  1. Something about new-style classes... `NewStyleInt16TestCase` is redundant on Python 3 I guess.  Can we indicate that somehow, so eventually it's clear we can delete it?
  1. Sorry, I think the new comments are largely inappropriate - at least, as comments in the source.  They would probably make a good labs blog post though.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [35916]) Branching to 'basic-py3-6025-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Addressed all review comments. I skipped the newstyle tests, added utility method for netstring formatting, and sanitized the comment about Python 3.

Started http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/basic-py3-6025-3

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. I guess there should be a .removal news fragment, for the removal of writing non-str.
  1. `LPTestCaseMixin.test_illegal` still has a subjective comment about Python 3, and some code that would be simplified by `iterbytes`.

That's all.  Please merge once those are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `closed`|
|-|-|

(In [35925]) Merge basic-py3-6025-3.

Author: itamar
Review: exarkun
Fixes: [#6025](https://github.com/twisted/twisted/issues/6025)

twisted.protocols.basic  now works on Python 3. Also you can't send arbitrary objects using NetstringReceiver.sendString (deprecated since v10.0).
